### DISCUSSION TITLE
How call the prefix and range of an IP-Address?

### DISCUSSION THREAD

**Discussion by martinum4** [netbox-community/netbox#8996] [Q&A]
Hello,
I'm on NetBox 3.1.8, but haven't noticed anything in the changelog that would mention something like this.

I noticed I can call the IPs of a Prefix via NBshell e.g. `Prefix.objects.get(id=1337).get_child_ips()` -> `RestrictedQuerySet [<IPAddress: 10.11.12.13/24>]`

However i noticed the reverse is not possible, there is no method to` get_parent_prefix` or similar, however on the website there is a nice generated table, so the information has to be obtainable _somehow_. kind of the same goes for IP-Ranges, i guess it borrows all the methods from the prefixes.

How do i accomplish this? 

Kind Regards

**Comment by candlerb** [comment]
Given that an "IP address" in Netbox also records the prefix length, you *could* just calculate it by masking with the prefix length.  e.g. if the address is "10.11.12.13/24" then the prefix (if it exists) should be "10.11.12.0/24".

Unfortunately, the prefix length which users enter manually is often wrong anyway :-(

You can instead locate ancestor prefixes by querying for all prefixes which contain the given IP address. This is how the web UI does it in the IP address view:
https://github.com/netbox-community/netbox/blob/v3.1.10/netbox/ipam/views.py#L631-L636

In nbshell:

```
>>> ip = IPAddress.objects.get(address="10.12.255.1/24")
>>> pfx = Prefix.objects.filter(vrf=ip.vrf, prefix__net_contains_or_equals=str(ip.address.ip))
>>> pfx
<PrefixQuerySet [<Prefix: 10.12.0.0/16>, <Prefix: 10.12.255.0/24>]>
```

If you sort this list by prefix length, and take the longest, then you have the nearest enclosing prefix object.  Similar logic is used in the script here:
https://github.com/netbox-community/reports/blob/master/reports/ipam-reports/ip-check-prefix.py
(except for efficiency this preloads all prefixes into RAM first)

> kind of the same goes for IP-Ranges, i guess it borrows all the methods from the prefixes.

I don't think it does, but again you can search as per [here](https://github.com/netbox-community/netbox/blob/v3.1.10/netbox/ipam/filtersets.py#L477-L486) to find ranges containing the given IP:

```
>>> import netaddr
>>> ip = netaddr.IPNetwork("10.12.255.245/24")
>>> IPRange.objects.filter(vrf=None, start_address__lte=ip, end_address__gte=ip)
<RestrictedQuerySet [<IPRange: 10.12.255.240-249/24>]>
```
